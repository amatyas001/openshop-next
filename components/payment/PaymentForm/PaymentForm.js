import { Box, Flex } from '@chakra-ui/core';
import { PAYMENT_FIELDS } from '@app/config';
import { Field, Spinner, PaymentFormControls } from '@app/components';
import { FieldShape } from '@app/lib/types';

/**
 * Displays a form which collecting billing details from the user amd
 * [PaymentFormControls](#paymentformcontrols) component to handle submit.
 *
 * @visibleName Payment Form
 * @example
 * ```jsx
 * <PaymentForm />
 * ```
 */
export const PaymentForm = (props) => {
  const { fields } = props;
  const [loading, setLoading] = React.useState(false);
  const [details, setDetails] = React.useState({
    address: '',
    email: '',
    name: '',
    phone: '',
  });

  return !loading ? (
    <Flex
      as='form'
      flexDirection='column'
      alignItems='center'
      mx='auto'
      width={{ sm: '100%', lg: '80%' }}
      {...props}
    >
      <Box
        as='fieldset'
        border='0'
        width='100%'
        mb='10px'
        borderBottom='1px'
        borderColor='gray.400'
      >
        {fields.map((item) => (
          <Field
            data-testid='input'
            key={item.name}
            label={item.name}
            id={item.name}
            type={item.type}
            placeholder={item.placeholder}
            required
            value={details[item.name]}
            onChange={(e) => {
              setDetails({ ...details, [item.name]: e.target.value });
            }}
          />
        ))}
      </Box>
      <PaymentFormControls setLoading={setLoading} details={details} />
    </Flex>
  ) : (
    <Spinner data-testid='form-spinner' text='Sending data...' />
  );
};

PaymentForm.defaultProps = {
  fields: PAYMENT_FIELDS,
};

PaymentForm.propTypes = {
  /**
   * Arryas of input field descriptions to generate [Fileds](#field)
   */
  fields: FieldShape,
};
