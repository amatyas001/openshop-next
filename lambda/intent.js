const axios = require('axios');
const stripe = require('stripe')(process.env.STRIPE_SECRET_KEY),
  headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
  };

exports.handler = async (event, context) => {
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
    };
  }

  const data = JSON.parse(event.body);

  if (data.cancel) {
    try {
      const intent = await stripe.paymentIntents.cancel(data.cancel, {
        cancellation_reason: 'requested_by_customer',
      });

      if (intent.error)
        return {
          statusCode: 400,
          headers,
          body: JSON.stringify({
            status: intent.error.message,
          }),
        };

      return {
        statusCode: 200,
        headers,
        body: JSON.stringify({
          id: intent.id,
        }),
      };
    } catch (err) {
      console.log(err);

      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({
          status: err,
        }),
      };
    }
  } else if (!data.items || !data.token) {
    return {
      statusCode: 400,
      headers,
      body: JSON.stringify({
        status: { message: 'Missing parameters!' },
      }),
    };
  } else {
    try {
      const store = await axios.get(
        //'https://openshop.netlify.app/storedata.json'
        'http://localhost:3000/storedata.json'
      );
      const amount = data.items.reduce((a, c) => {
        const purchased = store.data.items.find((item) => item.id === c.id);
        return a + purchased.price * 100;
      }, 0);
      const intent = await stripe.paymentIntents.create(
        {
          currency: 'usd',
          amount: amount,
          description: 'Order from store',
        },
        {
          idempotencyKey: data.token,
        }
      );

      return {
        statusCode: 200,
        headers,
        body: JSON.stringify({
          id: intent.id,
          secret: intent.client_secret,
        }),
      };
    } catch (err) {
      console.log(err);

      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({
          status: err,
        }),
      };
    }
  }
};
