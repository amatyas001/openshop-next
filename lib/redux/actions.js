export * from './cart';
export * from './filter';
export * from './panel';
export * from './payment';
