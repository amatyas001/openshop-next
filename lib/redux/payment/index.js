export * from './paymentCancel';
export * from './paymentCreate';
export * from './paymentStage';
