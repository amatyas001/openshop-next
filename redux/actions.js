export const ADD_TO_CART = 'ADD_TO_CART';
export const REMOVE_FROM_CART = 'REMOVE_FROM_CART';
export const CLEAR_CART = 'CLEAR_CART';
export const GET_AMOUNT = 'GET_AMOUNT';
export const CHECKOUT_PROGRESS = 'CHECKOUT_PROGRESS';
export const PAYMENT_PROGRESS = 'PAYMENT_PROGRESS';
export const PAYMENT_RESET = 'PAYMENT_RESET';

/**
 *  CART ACTIONS
 */
export function addToCart(item) {
  return {
    type: ADD_TO_CART,
    payload: item,
  };
}

export function removeFromCart(item) {
  return {
    type: REMOVE_FROM_CART,
    payload: item,
  };
}

export function clearCart() {
  return {
    type: CLEAR_CART,
  };
}

export function getAmount() {
  return {
    type: GET_AMOUNT,
  };
}

/**
 *  CHECKOUT FLOW
 */
export function checkoutReview() {
  return {
    type: CHECKOUT_PROGRESS,
    payload: { status: 'review' },
  };
}

export function checkoutDetails() {
  return {
    type: CHECKOUT_PROGRESS,
    payload: { status: 'details' },
  };
}

export function checkoutFinished() {
  return {
    type: CHECKOUT_PROGRESS,
    payload: { status: 'finished' },
  };
}

export function paymentActive() {
  return {
    type: PAYMENT_PROGRESS,
    payload: { status: true },
  };
}

export function paymentStart() {
  return {
    type: PAYMENT_PROGRESS,
    payload: { status: 'loading' },
  };
}

/**
 *  PAYMENT FLOW
 */
export function paymentConfirm(card, details, intent) {
  return {
    type: PAYMENT_PROGRESS,
    payload: {
      status: 'confirm',
      card,
      details,
      intent,
    },
  };
}

export function paymentSuccess(intent) {
  return {
    type: PAYMENT_PROGRESS,
    payload: {
      status: 'success',
      intent,
    },
  };
}

export function paymentCanceled(intent) {
  return {
    type: PAYMENT_PROGRESS,
    payload: {
      status: 'canceled',
      intent,
    },
  };
}

export function paymentError(error) {
  return {
    type: PAYMENT_PROGRESS,
    payload: { status: 'error', error },
  };
}

export function paymentReset() {
  return {
    type: PAYMENT_RESET,
  };
}

export function cancelPayment(id) {
  return async (dispatch) => {
    try {
      dispatch(paymentStart());
      const { loadStripe } = await import('@stripe/stripe-js');
      const stripe = await loadStripe(
        process.env.NEXT_PUBLIC_STRIPE_PUBLIC_KEY
      );
      stripe.paymentIntents.cancel(id, (err, intent) => {
        if (err) dispatch(paymentError(e.message));
        else dispatch(paymentCanceled(intent.id));
      });
    } catch (e) {
      dispatch(paymentError(e.message));
    }
  };
}

export function intentCart(card, details, items) {
  return async (dispatch) => {
    try {
      dispatch(paymentStart());
      const axios = await import('axios');
      const intent = await axios.post(
        process.env.NODE_ENV === 'production'
          ? 'https://openshop.netlify.com/.netlify/functions/intent'
          : 'http://localhost:9000/.netlify/functions/intent',
        {
          items,
        },
        {
          headers: { 'Content-Type': 'application/json' },
        }
      );
      dispatch(paymentConfirm(card, details, intent.data));
    } catch (e) {
      dispatch(paymentError(e.message));
    }
  };
}
