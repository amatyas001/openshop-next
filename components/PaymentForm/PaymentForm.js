import { CardElement, useStripe } from '@stripe/react-stripe-js';
import {
  Box,
  Flex,
  Heading,
  Link,
  SimpleGrid,
  Spinner,
  Text,
} from '@chakra-ui/core';
import { useDispatch, useSelector } from 'react-redux';
import { paymentIntent, paymentReview } from '../../redux/actions';
import { Button, Field } from '../index';

// Handle user data and send request to Stripe
export const PaymentForm = (props) => {
  const stripe = useStripe();
  const dispatch = useDispatch();
  const state = useSelector((state) => state);

  const [cardError, setCardError] = React.useState(null);
  const [cardComplete, setCardComplete] = React.useState(false);
  const [valid, setValid] = React.useState(false);
  const [details, setDetails] = React.useState({
    email: '',
    phone: '',
    name: '',
    address: '',
  });

  // validate form
  React.useEffect(() => {
    setValid(isFormValid(details, cardComplete));
  }, [details, cardComplete]);

  //validator
  const isFormValid = (items, card) => {
    let isValid = true;
    Object.values(items).forEach((item) => {
      if (item === '') isValid = false;
    });
    return card && isValid;
  };

  // submit handler
  const handleSubmit = async (e) => {
    if (!cardComplete) return;
    dispatch(paymentIntent(details, state.cart, state.payment.token));
  };

  return (
    <>
      {/* payment form */}
      <Flex as='form' flexDirection='column' alignItems='center' {...props}>
        <Box
          as='fieldset'
          border='0'
          width='100%'
          mb='10px'
          borderBottom='1px'
          borderColor='gray.400'
        >
          {/* name */}
          <Field
            label='name'
            id='name'
            type='text'
            placeholder='Jane Doe'
            required
            autoComplete='name'
            value={details.name}
            onChange={(e) => {
              setDetails({ ...details, name: e.target.value });
            }}
          />

          {/* address */}
          <Field
            label='address'
            id='address'
            type='text'
            placeholder='J191 S. Illinois Street, Everett, MA 02149'
            required
            autoComplete='address'
            value={details.address}
            onChange={(e) => {
              setDetails({ ...details, address: e.target.value });
            }}
          />

          {/* email */}
          <Field
            label='email'
            id='email'
            type='email'
            placeholder='janedoe@gmail.com'
            required
            autoComplete='email'
            value={details.email}
            onChange={(e) => {
              setDetails({ ...details, email: e.target.value });
            }}
          />

          {/* phone */}
          <Field
            label='phone'
            id='phone'
            type='tel'
            placeholder='(941) 555-0123'
            required
            autoComplete='tel'
            value={details.phone}
            onChange={(e) => {
              setDetails({ ...details, phone: e.target.value });
            }}
          />

          {/* card */}
          <Flex alignItems='center'>
            {/* label */}
            <Heading
              as='label'
              htmlFor='StripeElement'
              fontSize='1.5rem'
              alignSelf='start'
            >
              card
            </Heading>

            {/* test info */}
            <Text
              mt='-7px'
              mb='-7px'
              p='7px'
              fontSize='0.8rem'
              color='gray.600'
            >
              Use any date and CVC with <strong>4242 4242 4242 4242</strong>{' '}
              number for testing
            </Text>

            {/* error display */}
            <Text
              d={cardError ? 'inline-block' : 'none'}
              ml='auto'
              mt='-7px'
              mb='-7px'
              p='7px'
              fontSize='1rem'
              fontWeight='bold'
              bg='red.400'
              borderRadius='5px'
              color='gray.100'
            >
              {cardError && cardError.message}
            </Text>
          </Flex>

          {/* card element */}
          {stripe ? (
            <CardElement
              options={{
                hidePostalCode: true,
                iconStyle: 'solid',
                style: {
                  base: {
                    border: '1px solid black',
                    iconColor: '#1A202C',
                    color: '#1A202C',
                    fontWeight: 500,
                    fontFamily: 'Montserrat, Open Sans, Segoe UI, sans-serif',
                    fontSize: '16px',
                    fontSmoothing: 'antialiased',
                    ':-webkit-autofill': { color: '#fce883' },
                    '::placeholder': { color: '#4A5568' },
                  },
                  invalid: {
                    iconColor: '##C53030',
                    color: '#f#C53030',
                  },
                },
              }}
              onChange={(e) => {
                setCardError(e.error);
                setCardComplete(e.complete);
              }}
            />
          ) : (
            <Spinner size='20px' mx='auto' />
          )}
        </Box>

        {/* ToS info */}
        <Text fontStyle='italic' textTransform='uppercase' fontSize='0.8rem'>
          By proceeding you accept the{' '}
          <Link color='purple.800' fontWeight='bold'>
            Terms of Service
          </Link>{' '}
          of the company.
        </Text>

        {/* controls */}
        <SimpleGrid
          columns={{ sm: 1, md: 2 }}
          spacing='20px'
          width='80%'
          mb='15px'
        >
          {/* review button */}
          <Button
            bg='gray.600'
            color='gray.100'
            width='100%'
            onClick={() => dispatch(paymentReview())}
          >
            BACK TO REVIEW
          </Button>

          {/* confirm button */}
          <Button
            bg='purple.800'
            color='gray.100'
            width='100%'
            disabled={!valid}
            onClick={handleSubmit}
          >
            {!valid ? 'FILL THE FORM' : `PAY ${state.amount} $`}
          </Button>
        </SimpleGrid>
      </Flex>
    </>
  );
};
