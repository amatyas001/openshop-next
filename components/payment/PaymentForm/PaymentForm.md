```jsx
import { useDispatch } from 'react-redux';
import { paymentForm, addToCart, getAmount } from '@app/redux/actions';

const dispatch = useDispatch();

const item = {
  id: 0,
  img: '1.jpg',
  name: 'My fancy item',
  desc: 'The best item in the world',
  color: 'Rainbow',
  price: 100,
};

React.useEffect(() => {
  dispatch(addToCart(item));
  dispatch(getAmount());
  dispatch(paymentForm());
}, []);

<PaymentForm />;
```
