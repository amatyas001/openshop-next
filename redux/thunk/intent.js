import { paymentConfirm, paymentError } from '../actions';

// Creating payment intent through lambda function
export function paymentIntent(details, items, token) {
  return async (dispatch) => {
    try {
      const axios = await import('axios');
      /* istanbul ignore next */
      const intent = await axios.post(
        process.env.NODE_ENV === 'production'
          ? 'https://openshop.netlify.app/.netlify/functions/intent'
          : 'http://localhost:9000/.netlify/functions/intent',
        {
          // cart items
          items,
          // idempotency key
          token,
        },
        {
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );

      // push confirm screen
      dispatch(paymentConfirm(details, intent.data));
    } catch (error) {
      dispatch(paymentError(error));
    }
  };
}
