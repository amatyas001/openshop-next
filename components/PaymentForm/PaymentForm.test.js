import { create, act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as actions from '../../redux/thunk/intent';
import { PaymentForm } from './PaymentForm';
import { paymentReview } from '../../redux/actions';

const mockStore = configureStore([thunk]);

describe('<PaymentForm />', () => {
  let tree, store;

  const completeForm = async (form) => {
    let details = {};
    await act(async () => {
      await form.root
        .findByType('CardElement')
        .props.onChange({ complete: true, error: 'mock_error' });
      await form.root.findAllByType('input').map((item) => {
        details[item.props.id] = 'mock_value';
        item.props.onChange({ target: { value: 'mock_value' } });
      });
    });
    return details;
  };

  beforeEach(() => {
    store = mockStore({
      payment: { status: 'form', token: 'token' },
      cart: [{ id: 'item' }],
      amount: 100,
    });
    act(() => {
      tree = create(
        <Provider store={store}>
          <PaymentForm />
        </Provider>
      );
    });
  });

  it('should render without props', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render form element', () => {
    expect(tree.root.findByType('form')).toBeTruthy();
  });

  it('should render card element', () => {
    expect(tree.root.findByType('CardElement')).toBeTruthy();
  });

  it('should render enabled pay button when form complete', async () => {
    await completeForm(tree);
    expect(
      tree.root.findByProps({ 'data-testid': 'submit-button' }).props.disabled
    ).toEqual(false);
  });

  it('should render disabled pay button when form incomplete', () => {
    expect(
      tree.root.findByProps({ 'data-testid': 'submit-button' }).props.disabled
    ).toEqual(true);
  });

  it('should render back button', () => {
    expect(
      tree.root.findByProps({ 'data-testid': 'review-button' })
    ).toBeTruthy();
  });

  it('should send intent with details', async () => {
    const details = await completeForm(tree);
    const intent = jest
      .spyOn(actions, 'paymentIntent')
      .mockImplementation(() => jest.fn());
    await act(async () => {
      await tree.root
        .findByProps({ 'data-testid': 'submit-button' })
        .props.onClick({ preventDefault: jest.fn() });
    });
    expect(intent).toHaveBeenCalledTimes(1);
    expect(intent).toHaveBeenCalledWith(
      details,
      store.getState().cart,
      store.getState().payment.token
    );
  });

  it('should push to review state', () => {
    act(() => {
      tree.root.findByProps({ 'data-testid': 'review-button' }).props.onClick();
    });
    expect(store.getActions()).toEqual([paymentReview()]);
  });
});
