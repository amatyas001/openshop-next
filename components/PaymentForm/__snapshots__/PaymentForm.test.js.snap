// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaymentForm /> should render without props 1`] = `
<form
  aria-hidden={false}
  className="css-8o331q"
>
  <fieldset
    className="css-94sf3x"
  >
    <fieldset
      className="css-9x3bhi"
    >
      <label
        className="css-1xnp38i"
        htmlFor="name"
      >
        name
      </label>
      <input
        autoComplete="name"
        className="css-cth1sm"
        id="name"
        onChange={[Function]}
        placeholder="Jane Doe"
        required={true}
        type="text"
        value=""
      />
    </fieldset>
    <fieldset
      className="css-9x3bhi"
    >
      <label
        className="css-1xnp38i"
        htmlFor="address"
      >
        address
      </label>
      <input
        autoComplete="address"
        className="css-cth1sm"
        id="address"
        onChange={[Function]}
        placeholder="J191 S. Illinois Street, Everett, MA 02149"
        required={true}
        type="text"
        value=""
      />
    </fieldset>
    <fieldset
      className="css-9x3bhi"
    >
      <label
        className="css-1xnp38i"
        htmlFor="email"
      >
        email
      </label>
      <input
        autoComplete="email"
        className="css-cth1sm"
        id="email"
        onChange={[Function]}
        placeholder="janedoe@gmail.com"
        required={true}
        type="email"
        value=""
      />
    </fieldset>
    <fieldset
      className="css-9x3bhi"
    >
      <label
        className="css-1xnp38i"
        htmlFor="phone"
      >
        phone
      </label>
      <input
        autoComplete="tel"
        className="css-cth1sm"
        id="phone"
        onChange={[Function]}
        placeholder="(941) 555-0123"
        required={true}
        type="tel"
        value=""
      />
    </fieldset>
    <div
      className="css-70qvj9"
    >
      <label
        className="css-1gjn3mk"
        htmlFor="StripeElement"
      >
        card
      </label>
      <p
        className="css-1i3anel"
      >
        Use any date and CVC with 
        <strong>
          4242 4242 4242 4242
        </strong>
         
        number for testing
      </p>
      <p
        className="css-cz9gy2"
      />
    </div>
    <CardElement
      data-testid="cardElement"
      onChange={[Function]}
      options={
        Object {
          "hidePostalCode": true,
          "iconStyle": "solid",
          "style": Object {
            "base": Object {
              ":-webkit-autofill": Object {
                "color": "#fce883",
              },
              "::placeholder": Object {
                "color": "#4A5568",
              },
              "border": "1px solid black",
              "color": "#1A202C",
              "fontFamily": "Montserrat, Open Sans, Segoe UI, sans-serif",
              "fontSize": "16px",
              "fontSmoothing": "antialiased",
              "fontWeight": 500,
              "iconColor": "#1A202C",
            },
            "invalid": Object {
              "color": "#f#C53030",
              "iconColor": "##C53030",
            },
          },
        }
      }
    />
  </fieldset>
  <p
    className="css-1yab5hb"
  >
    By proceeding you accept the
     
    <a
      className="css-1c5zfe9"
    >
      Terms of Service
    </a>
     
    of the company.
  </p>
  <div
    className="css-1rsrz83"
  >
    <button
      className="css-1h13z07"
      data-cy="button"
      data-testid="review-button"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="button"
    >
      BACK TO REVIEW
    </button>
    <button
      className="css-187kagv"
      data-cy="button"
      data-testid="submit-button"
      disabled={true}
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
      type="submit"
    >
      FILL THE FORM
    </button>
  </div>
</form>
`;
